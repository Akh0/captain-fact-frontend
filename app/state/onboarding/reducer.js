import { Record, Map } from 'immutable'
import { createAction, handleActions } from 'redux-actions'

import OnboardingStep from './record'

export const addStep = createAction('ONBOARDING/ADD_STEP')
export const removeStep = createAction('ONBOARDING/REMOVE_STEP_BY_NAME')
export const disable = createAction('ONBOARDING/DISABLE')
export const enable = createAction('ONBOARDING/ENABLE')

const INITIAL_STATE = new Record({
  steps: new Map()
})

const OnboardingReducer = handleActions({
  [addStep]: (state, {payload: step}) => {
    return state.steps.has(step.uniqueId) ?
      state :
      state.update('steps', steps => steps.set(step.uniqueId, OnboardingStep(step)))
  }
}, INITIAL_STATE())

export default OnboardingReducer
