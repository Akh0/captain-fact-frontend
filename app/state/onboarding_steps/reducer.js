import { Record, List, Map } from 'immutable'

import OnboardingStep from './record'

export const addStep = createAction('ONBOARDING/ADD_STEP')
export const removeStep = createAction('ONBOARDING/REMOVE_STEP_BY_NAME')
export const disable = createAction('ONBOARDING/DISABLE')
export const enable = createAction('ONBOARDING/ENABLE')

const INITIAL_STATE = new Record({
  steps: new List(),
  showOnboarding: false
})

const OnboardingStepsReducer = handleActions({
  [disable]: (state, {payload}) => state.set('showOnboarding', false),
  [enable]: (state, {payload}) => state.set('showOnboarding', true),
  [addStep]: (state, {payload}) => {
    const step = OnboardingStep(payload)
    const steps = state.steps.insert(state.steps.length, step)
    return state.set('steps', steps)
  },
  [removeStep]: (state, {payload}) => state.set('showOnboarding', true)
})