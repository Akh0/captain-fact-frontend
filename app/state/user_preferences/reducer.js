import { Record } from 'immutable'
import { createAction, handleActions } from 'redux-actions'
import { MOBILE_WIDTH_THRESHOLD } from '../../constants'


export const toggleSidebar = createAction('USER_PREFERENCES/TOGGLE_SIDEBAR')
export const closeSidebar = createAction('USER_PREFERENCES/CLOSE_SIDEBAR')
export const changeLocale = createAction('USER_PREFERENCES/CHANGE_LOCALE')
export const changeVideosLanguageFilter = createAction('USER_PREFERENCES/CHANGE_VIDEOS_LANGUAGE_FILTER')
export const changeVideosPublisherFilter = createAction('USER_PREFERENCES/CHANGE_VIDEOS_PUBLISHER_FILTER')
export const toggleAutoscroll = createAction('STATEMENTS/TOGGLE_AUTOSCROLL')

export const browserLocale = () =>
  (window.navigator.userLanguage || window.navigator.language || 'en').split('-')[0].toLowerCase()

const isMobile = window.innerWidth <= MOBILE_WIDTH_THRESHOLD

const Preferences = new Record({
  // Disable autoscroll and sidebar expended by default on mobile
  sidebarExpended: !isMobile,
  locale: 'en',
  enableAutoscroll: !isMobile,
  videosLanguageFilter: null,
  videosPublisherFilter: 'All'
})

const loadState = () => {
  let localStoragePrefs = {}
  try {
    localStoragePrefs = JSON.parse(localStorage.preferences)
  } catch (e) {
    localStorage.preferences = JSON.stringify(Preferences())
    localStoragePrefs = {locale: browserLocale()}
  }
  return Preferences().merge(localStoragePrefs)
}

const updateState = (state, key, value) => {
  console.log('update ' + key + '  ' + value)
  state = state.set(key, value)
  localStorage.preferences = JSON.stringify(state.toJSON())
  console.log(state)
  return state
}

const UserPreferencesReducer = handleActions({
  [toggleSidebar]: state => updateState(state, 'sidebarExpended', !state.sidebarExpended),
  [closeSidebar]: state => updateState(state, 'sidebarExpended', false),
  [changeLocale]: (state, {payload}) => updateState(state, 'locale', payload),
  [changeVideosLanguageFilter]: (state, {payload}) => updateState(state, 'videosLanguageFilter', payload),
  [changeVideosPublisherFilter]: (state, {payload}) => updateState(state, 'videosPublisherFilter', payload),
  [toggleAutoscroll]: state => updateState(state, 'enableAutoscroll', !state.enableAutoscroll)
}, loadState())

export default UserPreferencesReducer