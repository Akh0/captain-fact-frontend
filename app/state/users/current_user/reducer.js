import { Record, Set } from 'immutable'
import { createAction, handleActions } from 'redux-actions'

import User from '../record'

// Actions
export const set = createAction('CURRENT_USER/SET')
export const userLogin = createAction('CURRENT_USER/LOGIN')
export const setLoading = createAction('CURRENT_USER/SET_LOADING')
export const setPosting = createAction('CURRENT_USER/SET_POSTING')
export const reset = createAction('CURRENT_USER/RESET')
export const completeOnboardingStep = createAction('CURRENT_USER/COMPLETE_ONBOARDING_STEP')
export const removeOnboardingStep = createAction('CURRENT_USER/REMOVE_ONBOARDING_STEP')
export const setOnboardingSteps = createAction('CURRENT_USER/SKIP_TOUR')

// Initial reducer state
const INITIAL_STATE = new Record({
  error: null,
  isLoading: false,
  isPosting: false,
  data: new User()
})

// Reducer
const CurrentUserReducer = handleActions({
  [set]: {
    next: (state, {payload}) => (
      state.mergeDeep({
        data: new User(payload),
        error: null,
        isLoading: false
      })
    ),
    throw: (state, {payload}) =>
      state.merge({error: payload, isLoading: false})
  },
  [userLogin]: {
    next: (state, {payload}) => {
      return state.mergeDeep({data: payload || {}, error: null, isPosting: false})
    },
    throw: (state, {payload}) => {
      return state.merge({error: payload, isPosting: false})
    }
  },
  [setLoading]: (state, {payload}) =>
    state.set('isLoading', payload),
  [setPosting]: (state, {payload}) =>
    state.set('isPosting', payload),
  [completeOnboardingStep]: (state, {payload}) =>
    state.updateIn(['data', 'completed_onboarding_steps'], completedSteps =>
      completedSteps.add(payload)
    ),
  [setOnboardingSteps]: (state, {payload}) =>
    state.setIn(['data', 'completed_onboarding_steps'], new Set(payload)),
  [removeOnboardingStep]: (state, {payload}) =>
    state.updateIn(['data', 'completed_onboarding_steps'], s => s.delete(payload)),
  [reset]: () =>
    INITIAL_STATE()
}, INITIAL_STATE())

export default CurrentUserReducer
