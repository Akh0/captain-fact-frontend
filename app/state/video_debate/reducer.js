import { combineReducers } from "redux"

import VideoReducer from './video/reducer'
import CommentsReducer from './comments/reducer'
import StatementsReducer  from './statements/reducer'
import VideoDebateHistoryReducer  from './history/reducer'


const VideoDebateReducer = combineReducers({
  video: VideoReducer,
  statements: StatementsReducer,
  comments: CommentsReducer,
  history: VideoDebateHistoryReducer
})
export default VideoDebateReducer
