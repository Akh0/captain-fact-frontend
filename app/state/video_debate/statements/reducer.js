import { Record, List } from 'immutable'
import { combineActions, createAction, handleActions } from 'redux-actions'
import { change, destroy } from 'redux-form'

import Statement from './record'


export const fetchStatements = createAction('STATEMENTS/FETCH')
export const setLoading = createAction('STATEMENTS/SET_LOADING')
export const setSubmitting = createAction('STATEMENTS/SET_SUBMITTING')
export const add = createAction('STATEMENTS/ADD')
export const update = createAction('STATEMENTS/UPDATE')
export const remove = createAction('STATEMENTS/REMOVE')
export const setScrollTo = createAction('STATEMENTS/SET_SCROLL_TO')
export const incrementFormCount = createAction('STATEMENTS/INCREMENT_FORM_COUNT', () => 1)
export const decrementFormCount = createAction('STATEMENTS/DECREMENT_FORM_COUNT', () => -1)

// Statement form actions
const STATEMENT_FORM_NAME = 'StatementForm'
export const changeStatementFormSpeaker = ({id}) => change(STATEMENT_FORM_NAME, 'speaker_id', id)
export const closeStatementForm = () => destroy(STATEMENT_FORM_NAME)


const INITIAL_STATE = new Record({
  isLoading: false,
  isSubmitting: false,
  errors: null,
  data: new List(),
  scrollTo: null,
  formsCount: 0
})

const StatementsReducer = handleActions({
  [fetchStatements]: {
    next: (state, {payload}) => state.merge({
      data: new List(payload.map(s => new Statement(s))).sortBy(st => st.time),
      isLoading: false
    }),
    throw: (state, {payload}) => state.merge({isLoading: false, errors: payload})
  },
  [setLoading]: (state, isLoading) =>
    state.set('isLoading', isLoading),
  [setSubmitting]: (state, {payload}) => state.set('isSubmitting', payload),
  [add]: (state, {payload}) => {
    const statementIdx = getInsertPosition(state.data, payload)
    const statement = new Statement(payload)
    return state.update('data', statements => statements.insert(statementIdx, statement))
  },
  [update]: (state, {payload}) => {
    const statementIdx = state.data.findIndex(s => s.id === payload.id)
    if (statementIdx !== -1)
      return state.update('data', data => {
        data = data.delete(statementIdx)
        const newStatementIdx = getInsertPosition(data, payload)
        return data.insert(newStatementIdx, new Statement(payload))
      })
    return state
  },
  [remove]: (state, {payload: {id}}) => {
    const statementIdx = state.data.findIndex(s => s.id === id)
    if (statementIdx !== -1)
      return state.update('data', data => data.delete(statementIdx))
    return state
  },
  [setScrollTo]: (state, {payload}) => state.set('scrollTo', payload),
  [combineActions(incrementFormCount, decrementFormCount)]: (state, {payload}) =>
    state.set('formsCount', state.formsCount + payload)
}, INITIAL_STATE())

function getInsertPosition(data, newStatement) {
  const position = data.findIndex(st => newStatement.time < st.time)
  return position !== -1 ? position : data.size
}

export default StatementsReducer