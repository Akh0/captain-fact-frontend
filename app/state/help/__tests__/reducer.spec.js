import reducer, { setLoading, reset } from '../reducer'

const INITIAL_STATE = reducer(undefined, {})

it('should return the initial state', () => {
  expect(INITIAL_STATE).toMatchSnapshot()
})

it('setLoading', () => {
  expect(reducer(INITIAL_STATE, setLoading(true))).toMatchSnapshot()
  expect(reducer(INITIAL_STATE, setLoading(false))).toMatchSnapshot()
})

it('reset', () => {
  const modifiedState = reducer(INITIAL_STATE, setLoading(true))
  expect(reducer(modifiedState, reset())).toMatchSnapshot()
})
