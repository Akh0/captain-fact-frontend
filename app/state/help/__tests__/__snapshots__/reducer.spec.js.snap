// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reset 1`] = `
Immutable.Record {
  "markdownContent": "",
  "isLoading": false,
  "error": null,
}
`;

exports[`reset 2`] = `
Immutable.Record {
  "markdownContent": "",
  "isLoading": true,
  "error": null,
}
`;

exports[`reset 3`] = `
Immutable.Record {
  "markdownContent": "",
  "isLoading": false,
  "error": null,
}
`;

exports[`setLoading 1`] = `
Immutable.Record {
  "markdownContent": "",
  "isLoading": false,
  "error": null,
}
`;

exports[`setLoading 2`] = `
Immutable.Record {
  "markdownContent": "",
  "isLoading": true,
  "error": null,
}
`;

exports[`setLoading 3`] = `
Immutable.Record {
  "markdownContent": "",
  "isLoading": false,
  "error": null,
}
`;

exports[`should return the initial state 1`] = `
Immutable.Record {
  "markdownContent": "",
  "isLoading": false,
  "error": null,
}
`;
